lookout/style/format/analyzer.py

```diff
@@ -290,6 +240,159 @@ def train(cls, ptr: ReferencePointer, config: Mapping[str, Any], data_service: D
         cls._log.info("trained %s", model)
         return model
 
+    def render_comment_text(
+            self, language: str, line_number: int, code_lines: List[str], new_code_line: str,
+            winners: List[int], vnodes: List[VirtualNode], fixed_labels: List[int],
+            confidence: int, feature_extractor: FeatureExtractor) -> str:
+        """
+        Generate the text of the comment at the specified line.
+
+        :param language: Programming language of the code.
+        :param line_number: Line number for the comment.
+        :param code_lines: Original file code lines.
+        :param new_code_line: Code line suggested by our model.
+        :param winners: Winner rule indices.
+        :param vnodes: Corrected VirtualNode-s on the line.
+        :param fixed_labels: Predicted labels for those `vnodes`.
+        :param confidence: Overall confidence in the prediction, 0-100.
+        :param feature_extractor: FeatureExtractor involved.
+        :return: string with the generated comment.
+        """
+        rules = self.model[language]
+        _describe_rule = functools.partial(describe_rule, feature_extractor=feature_extractor)
+        _describe_change = functools.partial(
+            get_change_description, feature_extractor=feature_extractor)
+        return self.comment_template.render(
+                config=self.config,                # configuration of the analyzer
+                language=language,                 # programming language of the code
+                line_number=line_number,           # line number for the comment
+                code_lines=code_lines,             # original file code lines
+                new_code_line=new_code_line,       # code line suggested by our model
+                rules=rules.rules,                 # list of rules belonging to the model
+                winners=winners,                   # winner rule indices
+                vnodes=vnodes,                     # corrected VirtualNode-s
+                fixed_labels=fixed_labels,         # predicted labels for those nodes
+                confidence=confidence,             # overall confidence in the prediction, 0-100
+                describe_rule=_describe_rule,      # function to format a rule as text
+                describe_change=_describe_change,  # function to format a change as text
+                zip=zip,                           # Jinja2 does not have zip() by default
```

Comment:
Should be removed if we don't use in the template. I guess it didn't work? Does it work as `zip(a, b, c)`?

Reply:
oh, wait a second...